controllers/progressivesync_controller.go

```diff
@@ -418,12 +312,135 @@ func (r *ProgressiveSyncReconciler) syncApp(ctx context.Context, appName string)
 	return r.ArgoCDAppClient.Sync(ctx, &syncReq)
 }
 
-// patchStatus patches the progressive sync object status
-func (r *ProgressiveSyncReconciler) patchStatus(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
-	key := client.ObjectKeyFromObject(pr)
-	latest := syncv1alpha1.ProgressiveSync{}
-	if err := r.Client.Get(ctx, key, &latest); err != nil {
-		return err
+// updateStatusWithRetry updates the progressive sync object status with backoff
+func (r *ProgressiveSyncReconciler) updateStatusWithRetry(ctx context.Context, pr *syncv1alpha1.ProgressiveSync) error {
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+
+		key := client.ObjectKeyFromObject(pr)
+		latest := syncv1alpha1.ProgressiveSync{}
+		if err := r.Client.Get(ctx, key, &latest); err != nil {
+			return err
+		}
+		latest.Status = pr.Status
+		if err := r.Client.Status().Update(ctx, &latest); err != nil {
+			return err
+		}
+		return nil
+
+	})
+	return retryErr
+}
+
+// reconcileStage reconcile a ProgressiveSyncStage
+func (r *ProgressiveSyncReconciler) reconcileStage(ctx context.Context, ps syncv1alpha1.ProgressiveSync, stage syncv1alpha1.ProgressiveSyncStage) (syncv1alpha1.ProgressiveSync, reconcile.Result, error, bool) {
```

Comment:
I'm not a fan of returning four different values, to be honest. I'm aware that it's idiomatic Go to return two values for error-handling, but having three other return values makes this difficult to follow. 

I think as a starting point, named return values (if at all possible?) and definitely adding documentation would make this bettre, as it's not immediately clear what the `bool` value means. 

Further refactoring to avoid 3 non-error return values would probably be beneficial, but I understand if we'd rather do that later. :slightly_smiling_face:

Reply:
Well, I just saw https://github.com/fluxcd/helm-controller/blob/e25d6894bfdc547101e60dbaf739f99ab4f005ef/controllers/helmrelease_controller.go#L184 that Seb linked and can see they have 3 returns values (2 + error) - so this isn't a 'first' :slightly_smiling_face: 

I'm still not overly fond of multiple return values, but at least naming and/or documenting them something would be a good first step. 

Truth is, the whole reconciliation loop has multiple responsibilities - or at least a wide scope of responsibility -, making it harder to have a single return value.

Reply:
I was able to remove a variable because we can use `result.Requeue` to see if we need to requeue or not.

I don't think we can remove the ProgressiveSync variable unless we start doing things with pointer, but we decided to follow a value semantic.

Are you happy with the latest implementation? Is there anything else do you think we can improve?

Reply:
Yea, looks good :+1: